### PR TITLE
switch to bili for CJS

### LINKED ISSUE #54: Move from ESM to bundled package
esm seems to make this package a lot bigger than it needs to be 

https://packagephobia.now.sh/result?p=parse-commit-message
https://bundlephobia.com/result?p=parse-commit-message@3.2.3

Would you be open to moving to a setup where the files are built and `esm` isn't a runtime dependancy?

I'm putting together a PR to move to https://github.com/developit/microbundle

or 

https://github.com/egoist/bili

### PR DESCRIPTION
**Pull request by hipstersmoothie**
having esm as a runtime dependency isn't the best situation. Instead `bili` will bundle the code for CJS without the added runtime complexity.

closes #54

### CODE REVIEW

**Comment by hipstersmoothie** [comment]
@tunnckoCore Is this something you are open to?

**Comment by aleclarson** [comment]
To clarify, we'll be using this library in [auto-release](https://github.com/intuit/auto-release) (soon to be renamed to just `auto`). 😊

**Comment by tunnckoCore** [comment]
I don't see how this can be a problem when we have the `module` field. And that's absolutely intentional. Both Rollup and Webpack understands it and uses it if it is found. 

In the Webpack case, as far as I know, you may need to put `'module'` before `'main'`.

I'm moving soon everything (most of my packages) to the https://github.com/tunnckoCore/monorepo, so in anyway I'll have Babel 7 + TypeScript pipeline. But still will follow that pattern `main: index.js` (esm bridge) for cjs and `module: dist/index.js` (compiled code)

<blockquote><div> GitHub</div><div><strong><a href="https://github.com/tunnckoCore/monorepo">tunnckoCore/monorepo</a></strong></div><div>Mono-repo including scoped and non-scoped packages, independently versioned. Managed by @Lerna and @YarnPKG. Using Jest, @ESLint, @Prettier and TypeScript through @Babel 7. Automatic dependencies u...</div></blockquote>

**Comment by tunnckoCore** [comment]
oops, accidentally clicked "Comment and close", sorry

**Comment by tunnckoCore** [comment]
So, what's the problem? I really liked that pattern.

**Comment by hipstersmoothie** [comment]
Sweet. That's basically what I'm looking for. 

The problem is I'm in node and import/export is still experimental in node an hidden under a flag. So not many user are using it.

My problem with esm is that it increases the bundle size of this tiny module considerably.

https://packagephobia.now.sh/result?p=parse-commit-message (71.7kB install for node)
https://bundlephobia.com/result?p=parse-commit-message@3.2.3 (5.3kb install size for modules supporting platforms)

<blockquote><div><strong><a href="https://packagephobia.now.sh/result?p=parse-commit-message">Package Phobia</a></strong></div><div>Find the cost of installing a node dependency</div></blockquote>
<blockquote><div><strong><a href="https://bundlephobia.com">parse-commit-message@3.2.3 ❘ BundlePhobia</a></strong></div><div>Find the performance impact of adding a npm package to your bundle.</div></blockquote>

**Comment by hipstersmoothie** [comment]
Some minifcation goes on with the second number, but it's still quite a big difference.

(I could be wrong though feel free to correct me)

**Comment by tunnckoCore** [comment]
The `esm` is behind the Node's flag anyway.

I understand, but still why you are looking so hard on bundle size when `auto` isn't an app (or I'm wrong?) and it's users won't care much. Also, 50-60kb are mostly nothing compared to what most of us have as devDependency - at least eslint + codestyle config (eslint + airbnb is around 9k packages) .

Don't get me wrong, I was on that team (no bundling, transpiling or etc stuff, just 2-3 devDeps for linting and testing, 100% coverage and CJS) until recently. Before few months started migrating to `esm` loader and that current pattern here. And all that was because I want future proof packages. If I never touch them they still will be working (one day even in the browsers natively, loading directly from npm? haha, why not), because they are using pure ES Modules and features existing in browsers and node since 8.10. And not forget that they will be available through unpkg.com both the CJS and ESM, so...

Just recently starting to push much harder to setup huge monorepo with all the cool things. And actually I always was thinking about something similar to monorepo, even before they existed. And I really feel good that finally everything is almost ready. The only problem with monorepos is that they mess with automatic versioning and releasing and publishing on CI. 

Btw, don't know still what exactly is `auto-release`, but I have a lot of works around that thing - most of them are at @tunnckoCoreLabs org and this one [@tunnckocore/release-cli](https://github.com/tunnckoCore/release-cli) which should move it to the @standard-release which I started before month or two.

**Comment by hipstersmoothie** [comment]
`auto` works for more around PR labels. I've found this is a better model for outside contribution. Strict commits can be a hinderance in many cases.

 A user wants a plugin to parse the conventional commit messages though so we've settled on this package.

**Comment by hipstersmoothie** [comment]
It creates github releases as well as:

- maintains a changelog
- publishes releases
- pluggable for multiple platforms
- a few other tools for PR interaction are bundled int

**Comment by hipstersmoothie** [comment]
>  The only problem with monorepos is that they mess with automatic versioning and releasing and publishing on CI.

Oh and `auto` works out of the box with monorepos 😎

**Comment by hipstersmoothie** [comment]
> Also, 50-60kb are mostly nothing compared to what most of us have as devDependency.

True, but there's a reason we got there. 😉 Why make the same mistake with a modern tool?

**Comment by tunnckoCore** [comment]
> Why make the same mistake with a modern tool?

I also thought about making it `peerDependency` from the beginning and say to users to install `esm` if they want CJS support. Btw, there will be `esm` package in user's node_modules for sure, so he won't feel the problem.
And I don't think it can be mistake since it will be native soon and everything will work. The only thing will be to remove the `esm` as a dependency. I think it's a lot better approach to have it by default and will make more users happy than using it as peerDep for other users that care about size.

> True, but there's a reason we got there.

Interesting, can I ask what? If you plan to somehow bundling it, they point to use the `module`.

File: package.json [hunk 1]
```diff
@@ -21,18 +22,18 @@
...
-    "asia": "^0.19.7"
+    "asia": "^0.19.7",
+    "bili": "^3.4.2"
   },
   "files": [
     "src",
     "index.js"
   ],
-  "main": "index.js",
+  "main": "dist/parse-commit-message.cjs.js",
```
Comment (tunnckoCore) [inline package.json:36]: It's not enough. All bundlers don't export "interop"-ed result. So CJS users will need to `require('...').default` which I really hate, and that's what I'm doing in the esm bridge index.js file.

**Comment by hipstersmoothie** [comment]
> And I don't think it can be mistake since it will be native soon and everything will work.

import/export has been under the experimental flag since node 8? It's been on the horizon for a while, so I'm not all that sure adoption will be here soon.

> make more users happy than using it as peerDep for other users that care about size.

I agree that it should be a peer deep for your non-node users.

> True, but there's a reason we got there.

What I meant by this was more of a response to "mostly nothing compared to what most of us have as devDependency". The only reason tools like eslint are so big is because they never cared about code size from the beginning.  I know users don't want to download 5mb of javascript to create a release. 50-60kb of cruft is a lot when you are trying to keep lean. Repeat it a few times and you are adding megabytes

> If you plan to somehow bundling it, they point to use the module.

hmm ive seen other packages contradict this.

File: package.json (hunk 1 above)
Comment (hipstersmoothie) [reply package.json:36]: I'll take a look at this tomorrow (it's 3am my time)

**Comment by just-be-dev** [comment]
~Just to round back up on this, it seems that when running `npx auto --help` will fail due in part to this packages import of `esm`. I created an issue there at https://github.com/intuit/auto/issues/244. I wonder if `npx` and `esm` just fundamentally clash. I've been looking for reported issues but I haven't found a lot.~ 

~If you want to go the bundling route I generally tend to rely on [developit/microbundle](https://github.com/developit/microbundle) for all my library bundling needs. One dependency, one bundle, little-to-no config.~

~In the meantime I'll continue to search for potential solutions.~

**Edit**: Evidently my debugging abililties go out the window at night. Please ignore. >_>

**Comment by aleclarson** [comment]
@zephraph Hmm, shouldn't Auto be using v4 of `parse-commit-message` (found [here](https://github.com/tunnckoCore/monorepo/tree/master/packages/parse-commit-message))? That version comes with a CommonJS bundle, I think.

**Comment by just-be-dev** [comment]
Oh, yeah, my bad. I traced it down to index.js line 4... I thought that was this repo, but the statement on the other repo [matches](https://github.com/tunnckoCore/monorepo/blob/a2bf13527da064404aa76b21893180520d01d6d4/packages/parse-commit-message/index.js#L4.) what I was seeing (`Error: Cannot find module './dist/cjs'`).

**Comment by tunnckoCore** [comment]
Agree. Dont have much to say. 

So, is this really an error with the beta-x tag?

I tried it and dont have this error. We shouldnt have err bbecause it now has cjs bundle on the main and esm on the module field.

**Pull request closed without merging**